magento/magento2
# UPS negotiated rate error blocks checkout

### Description
Magento v2.2.2

This PR fixes an error when the UPS API returns a `//RatingServiceSelectionResponse/RatedShipment/NegotiatedRates` value for some, but not all configured service **Allowed Methods** and completely blocks checkout as the customer can't select any shipping options.

**Screenshot from checkout/cart URL**
![screen shot 2018-03-22 at 9 32 50 pm](https://user-images.githubusercontent.com/14304240/37807462-f651c334-2e1b-11e8-95e3-c7db94a99e38.png)

Based upon what the UPS API returns, this may or may not be reproducible. I'll include the XML for the 3 configured services I was receiving from UPS at the bottom. 

I believe it's safer to check whether the `NegotiatedRates` node exists for the current service `$code` in the `$allowedMethods` array. Currently the code assumes that if:
```
$negotiatedArr = $xml->getXpath("//RatingServiceSelectionResponse/RatedShipment/NegotiatedRates");
```
isn't `empty` then the `NegotiatedRates` node exists for all services. But if a `NegotiatedRates` node exists for service code **01 - UPS Next Day Air**, but not for service code **03 - UPS Ground**, then the `$negotiatedActive` check will be `true` while processing service code **03**. But when: 
```
$cost = $shipElement->NegotiatedRates->NetSummaryCharges->GrandTotal->MonetaryValue;
```
is executed a **Trying to get property of non-object** error will occur causing a 500 (Internal Service Error) response.

![screen shot 2018-03-22 at 10 36 47 pm](https://user-images.githubusercontent.com/14304240/37808617-af1bdecc-2e21-11e8-9427-d3330af6fe4b.png)


### Fixed Issues (if relevant)
n/a

### Manual testing scenarios
1. Enable **Flat Rate** (default is enabled) and **UPS** options in `Stores > Configuration > Sales > Shipping Methods`
2. Set up UPS configuration (see screenshot below)
    * The **Ship to Specific Countries** option was set to only _United States_
3. Add some products to the cart
    * Go to the mini cart and select **View and Edit Cart**
    * Select the **Estimate Shipping and Tax** dropdown
    * Enter a valid state and zip combination
or
    * Go to the mini cart and select **Proceed to Checkout**
    * Enter a valid address

![oee test_admin_admin_system_config_edit_section_carriers](https://user-images.githubusercontent.com/14304240/37807927-5fc128bc-2e1e-11e8-8ac2-0ed093be6e17.png)

```xml

<RatedShipment>
	<Service>
		<Code>01</Code>
	</Service>
	<RatedShipmentWarning>Your invoice may vary from the displayed reference rates</RatedShipmentWarning>
	<BillingWeight>
		<UnitOfMeasurement>
			<Code>LBS</Code>
		</UnitOfMeasurement>
		<Weight>1.0</Weight>
	</BillingWeight>
	<TransportationCharges>
		<CurrencyCode>USD</CurrencyCode>
		<MonetaryValue>35.70</MonetaryValue>
	</TransportationCharges>
	<ServiceOptionsCharges>
		<CurrencyCode>USD</CurrencyCode>
		<MonetaryValue>0.00</MonetaryValue>
	</ServiceOptionsCharges>
	<TotalCharges>
		<CurrencyCode>USD</CurrencyCode>
		<MonetaryValue>35.70</MonetaryValue>
	</TotalCharges>
	<GuaranteedDaysToDelivery>1</GuaranteedDaysToDelivery>
	<ScheduledDeliveryTime>10:30 A.M.</ScheduledDeliveryTime>
	<RatedPackage>
		<TransportationCharges>
			<CurrencyCode>USD</CurrencyCode>
			<MonetaryValue>35.70</MonetaryValue>
		</TransportationCharges>
		<ServiceOptionsCharges>
			<CurrencyCode>USD</CurrencyCode>
			<MonetaryValue>0.00</MonetaryValue>
		</ServiceOptionsCharges>
		<TotalCharges>
			<CurrencyCode>USD</CurrencyCode>
			<MonetaryValue>35.70</MonetaryValue>
		</TotalCharges>
		<Weight>0.6</Weight>
		<BillingWeight>
			<UnitOfMeasurement>
				<Code>LBS</Code>
			</UnitOfMeasurement>
			<Weight>1.0</Weight>
		</BillingWeight>
	</RatedPackage>
	<NegotiatedRates>
		<NetSummaryCharges>
			<GrandTotal>
				<CurrencyCode>USD</CurrencyCode>
				<MonetaryValue>28.26</MonetaryValue>
			</GrandTotal>
		</NetSummaryCharges>
	</NegotiatedRates>
</RatedShipment>
```
```xml

<RatedShipment>
	<Service>
		<Code>02</Code>
	</Service>
	<RatedShipmentWarning>Your invoice may vary from the displayed reference rates</RatedShipmentWarning>
	<BillingWeight>
		<UnitOfMeasurement>
			<Code>LBS</Code>
		</UnitOfMeasurement>
		<Weight>1.0</Weight>
	</BillingWeight>
	<TransportationCharges>
		<CurrencyCode>USD</CurrencyCode>
		<MonetaryValue>24.20</MonetaryValue>
	</TransportationCharges>
	<ServiceOptionsCharges>
		<CurrencyCode>USD</CurrencyCode>
		<MonetaryValue>0.00</MonetaryValue>
	</ServiceOptionsCharges>
	<TotalCharges>
		<CurrencyCode>USD</CurrencyCode>
		<MonetaryValue>24.20</MonetaryValue>
	</TotalCharges>
	<GuaranteedDaysToDelivery>2</GuaranteedDaysToDelivery>
	<ScheduledDeliveryTime/>
	<RatedPackage>
		<TransportationCharges>
			<CurrencyCode>USD</CurrencyCode>
			<MonetaryValue>24.20</MonetaryValue>
		</TransportationCharges>
		<ServiceOptionsCharges>
			<CurrencyCode>USD</CurrencyCode>
			<MonetaryValue>0.00</MonetaryValue>
		</ServiceOptionsCharges>
		<TotalCharges>
			<CurrencyCode>USD</CurrencyCode>
			<MonetaryValue>24.20</MonetaryValue>
		</TotalCharges>
		<Weight>0.6</Weight>
		<BillingWeight>
			<UnitOfMeasurement>
				<Code>LBS</Code>
			</UnitOfMeasurement>
			<Weight>1.0</Weight>
		</BillingWeight>
	</RatedPackage>
	<NegotiatedRates>
		<NetSummaryCharges>
			<GrandTotal>
				<CurrencyCode>USD</CurrencyCode>
				<MonetaryValue>19.49</MonetaryValue>
			</GrandTotal>
		</NetSummaryCharges>
	</NegotiatedRates>
</RatedShipment>
```
```xml

<RatedShipment>
	<Service>
		<Code>03</Code>
	</Service>
	<RatedShipmentWarning>Your invoice may vary from the displayed reference rates</RatedShipmentWarning>
	<BillingWeight>
		<UnitOfMeasurement>
			<Code>LBS</Code>
		</UnitOfMeasurement>
		<Weight>1.0</Weight>
	</BillingWeight>
	<TransportationCharges>
		<CurrencyCode>USD</CurrencyCode>
		<MonetaryValue>9.43</MonetaryValue>
	</TransportationCharges>
	<ServiceOptionsCharges>
		<CurrencyCode>USD</CurrencyCode>
		<MonetaryValue>0.00</MonetaryValue>
	</ServiceOptionsCharges>
	<TotalCharges>
		<CurrencyCode>USD</CurrencyCode>
		<MonetaryValue>9.43</MonetaryValue>
	</TotalCharges>
	<GuaranteedDaysToDelivery/>
	<ScheduledDeliveryTime/>
	<RatedPackage>
		<TransportationCharges>
			<CurrencyCode>USD</CurrencyCode>
			<MonetaryValue>9.43</MonetaryValue>
		</TransportationCharges>
		<ServiceOptionsCharges>
			<CurrencyCode>USD</CurrencyCode>
			<MonetaryValue>0.00</MonetaryValue>
		</ServiceOptionsCharges>
		<TotalCharges>
			<CurrencyCode>USD</CurrencyCode>
			<MonetaryValue>9.43</MonetaryValue>
		</TotalCharges>
		<Weight>0.6</Weight>
		<BillingWeight>
			<UnitOfMeasurement>
				<Code>LBS</Code>
			</UnitOfMeasurement>
			<Weight>1.0</Weight>
		</BillingWeight>
	</RatedPackage>
</RatedShipment>
```

### Contribution checklist
